File: frontend/src/lib/schemas/prompt-templates.ts
import { z } from "zod";

export const promptTemplateKeySchema = z.enum([
  "answer_generation",
  "summarization",
  "comparison",
  "citation_validation",
  "agent_planning",
]);

export const promptTemplateVersionStateSchema = z.enum([
  "draft",
  "review",
  "published",
]);

export const promptTemplateVariableSchema = z.object({
  name: z
    .string()
    .trim()
    .min(1)
    .max(64)
    .regex(/^[A-Za-z_][A-Za-z0-9_]*$/),
  description: z.string().trim().max(1000).nullable().optional(),
  required: z.boolean().default(true),
  default: z.unknown().nullable().optional(),
});

export const promptTemplateVersionSchema = z.object({
  version_id: z.string().uuid(),
  prompt_template_id: z.string().uuid(),
  template_key: promptTemplateKeySchema,
  version_number: z.number().int().min(1),
  state: promptTemplateVersionStateSchema,
  is_active: z.boolean(),
  content: z.string(),
  variables: z.array(z.record(z.string(), z.unknown())),
  variable_schema: z.record(z.string(), z.unknown()),
  preview_context: z.record(z.string(), z.unknown()),
  change_note: z.string().nullable(),
  source_version_number: z.number().int().min(1).nullable(),
  created_by_id: z.string().uuid().nullable(),
  reviewed_by_id: z.string().uuid().nullable(),
  published_by_id: z.string().uuid().nullable(),
  reviewed_at: z.string().nullable(),
  published_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const promptTemplateSchema = z.object({
  prompt_template_id: z.string().uuid(),
  organization_id: z.string().uuid(),
  template_key: promptTemplateKeySchema,
  name: z.string(),
  description: z.string().nullable(),
  category: z.string(),
  latest_version_number: z.number().int().min(1),
  active_version_number: z.number().int().min(1).nullable(),
  active_version_id: z.string().uuid().nullable(),
  active_state: promptTemplateVersionStateSchema.nullable(),
  active_published_at: z.string().nullable(),
  eval_run_count: z.number().int().nonnegative(),
  created_by_id: z.string().uuid().nullable(),
  updated_by_id: z.string().uuid().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const promptTemplateListSchema = z.object({
  items: z.array(promptTemplateSchema),
  total: z.number().int().nonnegative(),
  limit: z.number().int().positive(),
  offset: z.number().int().nonnegative(),
});

export const promptTemplateVersionListSchema = z.object({
  prompt_template_id: z.string().uuid(),
  template_key: promptTemplateKeySchema,
  items: z.array(promptTemplateVersionSchema),
  total: z.number().int().nonnegative(),
});

export const promptTemplateEvalResultSchema = z.object({
  evaluation_run_id: z.string().uuid(),
  evaluation_set_id: z.string().uuid(),
  run_name: z.string().nullable(),
  status: z.string(),
  summary: z.record(z.string(), z.unknown()).nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  started_at: z.string().nullable(),
  completed_at: z.string().nullable(),
});

export const promptTemplateEvalResultListSchema = z.object({
  prompt_template_id: z.string().uuid(),
  template_key: promptTemplateKeySchema,
  version_number: z.number().int().min(1),
  items: z.array(promptTemplateEvalResultSchema),
  total: z.number().int().nonnegative(),
  limit: z.number().int().positive(),
  offset: z.number().int().nonnegative(),
});

export const promptTemplateDetailSchema = z.object({
  template: promptTemplateSchema,
  active_version: promptTemplateVersionSchema.nullable(),
  versions: promptTemplateVersionListSchema,
  eval_results: promptTemplateEvalResultListSchema.nullable(),
});

export const createPromptTemplateDraftSchema = z.object({
  source_version_number: z.number().int().min(1).nullable().optional(),
  change_note: z.string().trim().max(1000).nullable().optional(),
});

export const updatePromptTemplateVersionSchema = z.object({
  content: z.string().min(1).max(64_000).optional(),
  variables: z.array(promptTemplateVariableSchema).max(100).optional(),
  variable_schema: z.record(z.string(), z.unknown()).optional(),
  preview_context: z.record(z.string(), z.unknown()).optional(),
  change_note: z.string().trim().max(1000).nullable().optional(),
});

export const publishPromptTemplateVersionSchema = z.object({
  change_note: z.string().trim().max(1000).nullable().optional(),
});

export const rollbackPromptTemplateSchema = z.object({
  version_number: z.number().int().min(1),
  change_note: z.string().trim().max(1000).nullable().optional(),
});

export const previewPromptTemplateSchema = z.object({
  version_number: z.number().int().min(1).nullable().optional(),
  content: z.string().min(1).max(64_000).nullable().optional(),
  variables: z
    .array(promptTemplateVariableSchema)
    .max(100)
    .nullable()
    .optional(),
  variable_schema: z.record(z.string(), z.unknown()).nullable().optional(),
  context: z.record(z.string(), z.unknown()).default({}),
});

export const promptTemplatePreviewResponseSchema = z.object({
  template_key: promptTemplateKeySchema,
  version_number: z.number().int().min(1).nullable(),
  rendered_prompt: z.string(),
  context: z.record(z.string(), z.unknown()),
});

export type PromptTemplateKey = z.infer<typeof promptTemplateKeySchema>;
export type PromptTemplateVersionState = z.infer<
  typeof promptTemplateVersionStateSchema
>;
export type PromptTemplateVariable = z.infer<
  typeof promptTemplateVariableSchema
>;
export type PromptTemplate = z.infer<typeof promptTemplateSchema>;
export type PromptTemplateVersion = z.infer<typeof promptTemplateVersionSchema>;
export type PromptTemplateList = z.infer<typeof promptTemplateListSchema>;
export type PromptTemplateDetail = z.infer<typeof promptTemplateDetailSchema>;
export type PromptTemplateEvalResultList = z.infer<
  typeof promptTemplateEvalResultListSchema
>;
export type CreatePromptTemplateDraftRequest = z.infer<
  typeof createPromptTemplateDraftSchema
>;
export type UpdatePromptTemplateVersionRequest = z.infer<
  typeof updatePromptTemplateVersionSchema
>;
export type PublishPromptTemplateVersionRequest = z.infer<
  typeof publishPromptTemplateVersionSchema
>;
export type RollbackPromptTemplateRequest = z.infer<
  typeof rollbackPromptTemplateSchema
>;
export type PreviewPromptTemplateRequest = z.input<
  typeof previewPromptTemplateSchema
>;
export type PromptTemplatePreview = z.infer<
  typeof promptTemplatePreviewResponseSchema
>;

File: backend/app/domains/prompt_templates/schemas/prompt_templates.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from app.models.enums import PromptTemplateKey, PromptTemplateVersionState

PromptTemplateKeyLiteral = Literal[
    "answer_generation",
    "summarization",
    "comparison",
    "citation_validation",
    "agent_planning",
]
PromptTemplateStateLiteral = Literal["draft", "review", "published"]


class PromptTemplateVariableDefinition(BaseModel):
    name: str = Field(min_length=1, max_length=64, pattern=r"^[A-Za-z_][A-Za-z0-9_]*$")
    description: str | None = Field(default=None, max_length=1000)
    required: bool = True
    default: Any | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        return value.strip()

    @field_validator("description")
    @classmethod
    def validate_description(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class CreatePromptTemplateDraftRequest(BaseModel):
    source_version_number: int | None = Field(default=None, ge=1)
    change_note: str | None = Field(default=None, max_length=1000)

    @field_validator("change_note")
    @classmethod
    def validate_change_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class UpdatePromptTemplateVersionRequest(BaseModel):
    content: str | None = Field(default=None, min_length=1, max_length=64_000)
    variables: list[PromptTemplateVariableDefinition] | None = Field(
        default=None,
        max_length=100,
    )
    variable_schema: dict[str, Any] | None = Field(default=None)
    preview_context: dict[str, Any] | None = Field(default=None)
    change_note: str | None = Field(default=None, max_length=1000)

    @field_validator("content")
    @classmethod
    def validate_content(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("content must not be blank")
        return value

    @field_validator("change_note")
    @classmethod
    def validate_change_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None

    @model_validator(mode="after")
    def validate_update_has_changes(self) -> UpdatePromptTemplateVersionRequest:
        if (
            self.content is None
            and self.variables is None
            and self.variable_schema is None
            and self.preview_context is None
            and self.change_note is None
        ):
            raise ValueError("At least one update field is required")
        return self


class PublishPromptTemplateVersionRequest(BaseModel):
    change_note: str | None = Field(default=None, max_length=1000)

    @field_validator("change_note")
    @classmethod
    def validate_change_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class RollbackPromptTemplateRequest(BaseModel):
    version_number: int = Field(ge=1)
    change_note: str | None = Field(default=None, max_length=1000)

    @field_validator("change_note")
    @classmethod
    def validate_change_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        return trimmed or None


class PromptTemplatePreviewRequest(BaseModel):
    version_number: int | None = Field(default=None, ge=1)
    content: str | None = Field(default=None, min_length=1, max_length=64_000)
    variables: list[PromptTemplateVariableDefinition] | None = Field(
        default=None,
        max_length=100,
    )
    variable_schema: dict[str, Any] | None = Field(default=None)
    context: dict[str, Any] = Field(default_factory=dict)

    @field_validator("content")
    @classmethod
    def validate_content(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not value.strip():
            raise ValueError("content must not be blank")
        return value


class PromptTemplateVersionResponse(BaseModel):
    version_id: str
    prompt_template_id: str
    template_key: PromptTemplateKeyLiteral
    version_number: int
    state: PromptTemplateStateLiteral
    is_active: bool
    content: str
    variables: list[dict[str, Any]]
    variable_schema: dict[str, Any]
    preview_context: dict[str, Any]
    change_note: str | None = None
    source_version_number: int | None = None
    created_by_id: str | None = None
    reviewed_by_id: str | None = None
    published_by_id: str | None = None
    reviewed_at: datetime | None = None
    published_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class PromptTemplateResponse(BaseModel):
    prompt_template_id: str
    organization_id: str
    template_key: PromptTemplateKeyLiteral
    name: str
    description: str | None = None
    category: str
    latest_version_number: int
    active_version_number: int | None = None
    active_version_id: str | None = None
    active_state: PromptTemplateStateLiteral | None = None
    active_published_at: datetime | None = None
    eval_run_count: int = 0
    created_by_id: str | None = None
    updated_by_id: str | None = None
    created_at: datetime
    updated_at: datetime


class PromptTemplateListResponse(BaseModel):
    items: list[PromptTemplateResponse]
    total: int
    limit: int
    offset: int


class PromptTemplateVersionListResponse(BaseModel):
    prompt_template_id: str
    template_key: PromptTemplateKeyLiteral
    items: list[PromptTemplateVersionResponse]
    total: int


class PromptTemplateEvalResultResponse(BaseModel):
    evaluation_run_id: str
    evaluation_set_id: str
    run_name: str | None = None
    status: str
    summary: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None


class PromptTemplateEvalResultListResponse(BaseModel):
    prompt_template_id: str
    template_key: PromptTemplateKeyLiteral
    version_number: int
    items: list[PromptTemplateEvalResultResponse]
    total: int
    limit: int
    offset: int


class PromptTemplateDetailResponse(BaseModel):
    template: PromptTemplateResponse
    active_version: PromptTemplateVersionResponse | None = None
    versions: PromptTemplateVersionListResponse
    eval_results: PromptTemplateEvalResultListResponse | None = None


class PromptTemplatePreviewResponse(BaseModel):
    template_key: PromptTemplateKeyLiteral
    version_number: int | None = None
    rendered_prompt: str
    context: dict[str, Any]


def validate_prompt_template_key(value: str) -> str:
    normalized = value.strip()
    if normalized not in {item.value for item in PromptTemplateKey}:
        raise ValueError("Unsupported prompt template key")
    return normalized


def validate_prompt_template_state(value: str) -> str:
    normalized = value.strip()
    if normalized not in {item.value for item in PromptTemplateVersionState}:
        raise ValueError("Unsupported prompt template state")
    return normalized
